File: algorithms/src/strings/DeleteSmiles.java
package strings;

import javax.annotation.Generated;


public class DeleteSmiles {

    @Generated(value = {}, comments = "Deletes smiles like :-))) or :-(( in a string")
    private StringBuilder deleteSmiles(String a, int length) {
        StringBuilder result = new StringBuilder();

        @Generated("a")
        int i = 0;

        while (i < length) {
            @Generated(comments = "Consume char 1 of pattern", value = {"#FFFFF0", "FFFFD0"}) int $1;
            char c1 = a.charAt(i++);
            if (i >= length || c1 != ':') {
                @Generated(value = {"#FFFFF0", "FFD0D0"}) int $m;
                result.append(c1);
                continue;
            }
            int $1$;
            int _1;

            @Generated(comments = "Consume char 2 of pattern", value = {"#F0FFF0", "D0FFD0"}) int $2;
            char c2 = a.charAt(i++);
            if (i >= length || c2 != '-') {
                @Generated(value = {"#F0FFF0", "FFD0D0"}) int $m;
                result.append(c1);
                result.append(c2);
                continue;
            }
            int $2$;
            int _2;

            @Generated(comments = "Consume char 3 of pattern", value = {"#F0F0FF", "D0D0FF"}) int $3;
            char c3 = a.charAt(i++);
            if (c3 != ')' && c3 != '(') {
                @Generated(value = {"#F0F0FF", "FFD0D0"}) int $m;
                result.append(c1);
                result.append(c2);
                result.append(c3);
                continue;
            }
            int $3$;
            int _3;

            @Generated(comments = "Consume repetitions of char 3", value = {"#F4F0FF", "E0D0FF"}) int $4;
            while (true) {
                if (i >= length) {
                    return result;
                }

                char c = a.charAt(i);
                if (c != c3) {
                    @Generated(value = {"#F4F0FF", "FFD0D0"}) int $m;
                    break;
                }
                i = i + 1;    // <----------------------- advance only if match; otherwise will re-parse from c4 position
            }
        }
        return result;
    }

    StringBuilder result = deleteSmiles("Hi:-)) :-)_:-(( :)", 18);
}

File: v1/algo/delete_smiles.js
test = function() {

    const deleteSmiles = vm.functionDeclaration(
        'deleteSmiles',
        [vm.variable('a'), vm.variable('length')],
        vm.sequenceStatement([
            vm.assignment(vm.varWrite('result'), vm.arrayLiteral([])),
            vm.assignment(vm.varWrite('i', ['a']), vm.number(0)),
            vm.whileStatement(
                vm.expression(vm.lt(), vm.variable('i'), vm.variable('length')),
                vm.sequenceStatement([
                    vm.group('// Consume char 1 of pattern','#FFFFF0', 'FFFFD0',
                        [                    vm.assignment(vm.varWrite('c1'), vm.arrItem('a', vm.varPostOp('i', true))),
                            vm.ifStatement(
                                vm.expression(vm.or(), vm.expression(vm.ge(), vm.variable('i'), vm.variable('length')), vm.expression(vm.ne(), vm.variable('c1'), vm.char(':'))),
                                vm.sequenceStatement([
                                    vm.group(undefined,'#FFFFF0', 'FFD0D0',
                                        [                                vm.assignment(undefined, vm.functionCall('push', [vm.variable('c1')], 'result')),
                                            vm.continueStatement()
                                        ])
                                ])
                            )
                        ]),
                    vm.lineComment(),
                    vm.group('// Consume char 2 of pattern','#F0FFF0', 'D0FFD0',
                        [                    vm.assignment(vm.varWrite('c2'), vm.arrItem('a', vm.varPostOp('i', true))),
                            vm.ifStatement(
                                vm.expression(vm.or(), vm.expression(vm.ge(), vm.variable('i'), vm.variable('length')), vm.expression(vm.ne(), vm.variable('c2'), vm.char('-'))),
                                vm.sequenceStatement([
                                    vm.group(undefined,'#F0FFF0', 'FFD0D0',
                                        [                                vm.assignment(undefined, vm.functionCall('push', [vm.variable('c1')], 'result')),
                                            vm.assignment(undefined, vm.functionCall('push', [vm.variable('c2')], 'result')),
                                            vm.continueStatement()
                                        ])
                                ])
                            )
                        ]),
                    vm.lineComment(),
                    vm.group('// Consume char 3 of pattern','#F0F8FF', 'A0D0FF',
                        [                    vm.assignment(vm.varWrite('c3'), vm.arrItem('a', vm.varPostOp('i', true))),
                            vm.ifStatement(
                                vm.expression(vm.and(), vm.expression(vm.ne(), vm.variable('c3'), vm.char(')')), vm.expression(vm.ne(), vm.variable('c3'), vm.char('('))),
                                vm.sequenceStatement([
                                    vm.group(undefined,'#F0F8FF', 'FFD0D0',
                                        [                                vm.assignment(undefined, vm.functionCall('push', [vm.variable('c1')], 'result')),
                                            vm.assignment(undefined, vm.functionCall('push', [vm.variable('c2')], 'result')),
                                            vm.assignment(undefined, vm.functionCall('push', [vm.variable('c3')], 'result')),
                                            vm.continueStatement()
                                        ])
                                ])
                            )
                        ]),
                    vm.lineComment(),
                    vm.group('// Consume repetitions of char 3','#F4F0FF', 'E0D0FF',
                        [                    vm.whileStatement(
                            vm.bool(true),
                            vm.sequenceStatement([
                                vm.ifStatement(
                                    vm.expression(vm.ge(), vm.variable('i'), vm.variable('length')),
                                    vm.sequenceStatement([
                                        vm.returnStatement(vm.variable('result'))
                                    ])
                                ),
                                vm.assignment(vm.varWrite('c'), vm.arrItem('a', vm.variable('i'))),
                                vm.ifStatement(
                                    vm.expression(vm.ne(), vm.variable('c'), vm.variable('c3')),
                                    vm.sequenceStatement([
                                        vm.group(undefined,'#F4F0FF', 'FFD0D0',
                                            [                                        vm.breakStatement()
                                            ])
                                    ])
                                ),
                                vm.assignment(vm.varWrite('i'), vm.expression(vm.plus(), vm.variable('i'), vm.number(1)))
                            ])
                        )
                        ])
                ])
            ),
            vm.returnStatement(vm.variable('result'))
        ]),
        '// Deletes smiles like :-))) or :-(( in a string'
    );

    const usage = vm.assignment(vm.varWrite('result'), vm.functionCall(deleteSmiles, [vm.string('Hi:-)) :-)_:-(( :)'), vm.number(18)]));

    return {
        code: vm.codeBlocks([deleteSmiles, usage]),
        entry: usage
    };
}();
